### ISSUE TITLE
Spring Cloud Stream App doesn't automatically shutdown

### ISSUE THREAD

**Issue by mminella** [spring-cloud/spring-cloud-stream#462]
When creating a Spring Cloud Task app that produces output on an SCSt channel, once the task is complete, the app does not shut down.  Adding `spring.main.web-environment=false` prevented Tomcat from starting but still does not allow the app to shut down.

**Comment by dsyer** [comment]
IMO we should expect that short lived apps might want to expose web endpoints anyway, so webEnvironment=true is perfectly natural. This is related to the issue I already raised about the lifecycle: if the task is only complete when the context is closed that doesn't leave much room for apps to listen on a socket for control plane events. The two events should be decoupled.

**Comment by dsyer** [comment]
Link to lifecycle issue: https://github.com/spring-cloud/spring-cloud-task/issues/102. FWIW I don't think a stream app should shut down after the context has started ever (if needs to listen for incoming messages), and that is independent of whether it has web (actuator) endpoints as well.

**Comment by mminella** [comment]
A few responses:
* +1 on the idea that the `webEnvironment=true` should work on a short lived application.  I'm not stating that it shouldn't, only that I tried it as a solution and it didn't work.
* This issue isn't related to your issue.  If I have a Stream application that is a source only.  The addition of the code to activate a source shouldn't impact the lifecycle of my application.  In other words, if I have an application that does shut down automatically once my processing is complete, adding a source to it shouldn't change that behavior (it currently does).

**Comment by mbogoevici** [comment]
The canonical test for this would be an application that sends a few messages from a CommandLineRunner and stops (should stop I suppose). My finger is currently pointing at the default poller, which we don't need for this type of scenario anyway. Will check once getting closer to a dev environment ...

**Comment by ilayaperumalg** [comment]
> if I have an application that does shut down automatically once my processing is complete, adding a source to it shouldn't change that behavior (it currently does).

Apart from the embedded servlet container threads, there will also be the `taskScheduler` threads created out of `@EnableIntegration` defined in `@EnableBinding` running on the application.

**Comment by ilayaperumalg** [comment]
One possible way is to shutdown the app using the `/shutdown` endpoint explicitly after the `CommandLineRunner` completes its job:

```
RestTemplate restTemplate = new RestTemplate(new SimpleClientHttpRequestFactory());
					try {
						// resolve port and management context path
						restTemplate.postForObject("http://localhost:8080" + "/shutdown", Object.class,
								Object.class);
					}
					catch (RestClientException e) {
						throw new RuntimeException(e.getMessage());
					}
```

**Comment by artembilan** [comment]
I moved a bit further. It doesn't start even without Spring Integration!
````java
@RefreshScope
@Configuration
public class MyConfig implements SmartLifecycle {

	@Override
	public boolean isAutoStartup() {
		return true;
	}

	@Override
	public void start() {
		System.out.println("STARTED");
	}

	@Override
	public void stop() {

	}

	@Override
	public void stop(Runnable callback) {

	}

	@Override
	public boolean isRunning() {
		return false;
	}

	@Override
	public int getPhase() {
		return 0;
	}

}
````

I wonder if it is know issue in Spring Cloud.
What I have noticed from the Spring Integration perspective that the `@Configuration` bean comes to the `MessagingAnnotationPostProcessor` later, than `DefaultLifecycleProcessor.start()` does its stuff.

So, it doesn't relate to Spring Integration, neither to Spring Cloud Stream.
That is something around `@Scope` logic.

CC @dsyer

**Comment by artembilan** [comment]
Yep!

We instantiate that bean only here `RefreshScope`:
````java
@EventListener
public void start(ContextRefreshedEvent event) {
	if (event.getApplicationContext() == this.context) {
		if (this.eager && this.registry != null) {
			for (String name : this.context.getBeanDefinitionNames()) {
				BeanDefinition definition = this.registry.getBeanDefinition(name);
				if (this.getName().equals(definition.getScope())
						&& !definition.isLazyInit()) {
					this.context.getBean(name);
				}
			}
		}
	}
}
````
What happens already after the ` DefaultLifecycleProcessor.start()` :smile:

**Comment by artembilan** [comment]
Sorry, guys, for posting to the wrong place.
My bad.

Thank you, @ilayaperumalg , for letting me to know!
:smile:

**Comment by mbogoevici** [comment]
If we wanted to shut down the application explicitly, a `context.close()` would suffice - we don't need to call into the actuator endpoints. That being said, I don't think this is something we can fix from within SCSt itself, as it's coming from Spring Integration support directly (i.e. the `taskScheduler` thread pool). We could eventually ensure that no thread pool is created, but that somehow feels like fighting the framework. Also, it doesn't solve the problem of web support. Overall, I agree with @dsyer that it might be better if something like https://github.com/spring-cloud/spring-cloud-task/issues/102 was in place, and that triggered the context close, rather than assuming that the application will shut down automatically. In addition to that, there might be other integrations that inadvertently create thread pools, so having a specific task-related event that triggers the context close will help address that too.

**Comment by mminella** [comment]
@dsyer can feel free to contradict this, but I haven't interpreted that spring-cloud/spring-cloud-task#102 would prevent the context from shutting down, just that if you had an app that you wanted to live longer, you could do so and have the task complete.  That's not what's at issue here.  In fact, it's the opposite IMHO.  

Fundamentally this becomes a bigger issue in that we can't have tasks emit events (either raw task events or batch events like we had in XD) and support the promised behavior of tasks in that they shut down on their own.  AKA, we're saying that we can't have true task based functionality (short lived microservices) that emit events via Spring Cloud Streams.  Obviously the lack of that type of functionality then also means that our orchestration story cannot come full circle.

**Comment by mbogoevici** [comment]
The motivation is opposite, but what they have in common AFAICT is that you can't have an active context and a complete task at the same time, unless @dsyer was suggesting to keep the application up with a closed context, as you mentioned.

**Comment by mminella** [comment]
Closing and addressing via a fix that will address this and spring-cloud/spring-cloud-task#102.

**Comment by steven-chou** [comment]
Hi, I know this issue has been closed for a while. But I just want to know if the following statement is true and aligned with the goal of this project.

For a short-lived micro-service or a function as service like AWS Lambda, even if I ONLY use the MessageChannel API provided by Spring Cloud Stream to produce the message(No message consumer function used), the service/function will never close/end as expected.

I've tried with the Spring Cloud Stream Kinesis binder (1.2.0) and with Spring Boot 2.1.7 and the app just won't shutdown unless I bring in the Spring Cloud Task and enable the close context or manually close the context myself.

If this is really how Spring Cloud Stream should work, I think this side effect should be documented.
